README.md
# ska

Lets you easily sign data, using symmetric-key algorithm encryption. Allows
you to validate signed data and identify possible validation errors. Uses
sha/hmac for signature encryption. Comes with shortcut functions for signing (and
validating) dictionaries.

[![NPM Version](https://img.shields.io/npm/v/skajs.svg)](https://www.npmjs.com/package/skajs)
[![Supported NodeJS versions](https://img.shields.io/node/v/skajs.svg)](https://www.npmjs.com/package/skajs)
[![Build Status](https://github.com/barseghyanartur/skajs/actions/workflows/node.js.yml/badge.svg)](https://github.com/barseghyanartur/skajs/actions)
[![Coverage Status](https://coveralls.io/repos/github/barseghyanartur/skajs/badge.svg?branch=main)](https://coveralls.io/github/barseghyanartur/skajs?branch=main)
[![License](https://img.shields.io/badge/license-MIT-blue.svg)](https://github.com/barseghyanartur/skajs/#License)

## Key concepts

Hosts, that communicate with each other, share the Secret Key, which is used
to sign data (requests). Secret key is never sent around.

One of the cases is signing of HTTP requests. Each (HTTP) request is signed
on the sender side using the shared Secret Key and as an outcome produces the
triple (`signature`, `auth_user`, `valid_until`) which are used to sign
the requests.

-   `signature` (`string`): Signature generated.
-   `auth_user` (`string`): User making the request. Can be anything.
-   `valid_until` (`float` or `string`): Signature expiration time (Unix timestamp).

On the recipient side, (HTTP request) data is validated using the shared
Secret Key. It's being checked whether signature is valid and not expired.

```
    ┌─────────────┐           Data              ┌─────────────┐
    │   Host 1    ├────────────────────────────>│   Host 2    │
    │ ─────────── │                             │ ─────────── │
    │ secret key  │                             │ secret key  │
    │ 'my-secret' │<────────────────────────────┤ 'my-secret' │
    └─────────────┘           Data              └─────────────┘
```

## Features

-   Sign URLs.
-   Sign dictionaries.
-   Validate signed dictionaries.

## Eco-system

Need ``ska`` for other languages? Check the following affiliated projects:

- [ska](https://github.com/barseghyanartur/ska): ``ska`` implementation
  for Python. This was the first implementation from which current project
  originated.
- [skaphp](https://github.com/barseghyanartur/skaphp): ``ska`` implementation
  for PHP (>= 7.2).

Generated signatures are intercompatible between Python, NodeJS and PHP
implementations.

## Installation

Latest stable version from NPM registry:

```shell
npm install skajs
```

## Usage examples

Usage example are present for both CommonJS and ESM.

### CommonJS

```shell
node examples.js
```

### ESM

```shell
node examples.mjs
```

### Basic usage

#### Sender side

Signing dictionaries and URLs is as simple as follows.

##### Required imports.

**CommonJS**

```javascript
const { signatureToDict, signURL } = require("skajs");
```

**ESM**

```javascript
import { signatureToDict, signURL } from "skajs";
```

##### Sign data

**Sample usage, sign a dictionary:**

```javascript
const signatureDict = signatureToDict("user", "your-secret_key");
```

**Sample output:**

```javascript
{
  signature: 'sf40lBWO5CquFfHr6jSXxhl2oW0=',
  auth_user: 'user',
  valid_until: '1631827551.6',
  extra: ''
}
```

**Adding of additional data to the signature works in the same way:**

```javascript
const signatureDict = signatureToDict(
    "user",
    "your-secret_key",
    {
        "email": "john.doe@mail.example.com",
        "first_name": "John",
        "last_name": "Doe",
    }
);
```

**Sample output:**

```javascript
{
  signature: 'B0sscS+xXWU+NR+9dBCoGFnDtlw=',
  auth_user: 'user',
  valid_until: '1631827551.6',
  extra: 'email,first_name,last_name',
  email: 'john.doe@mail.example.com',
  first_name: 'John',
  last_name: 'Doe',
}
```

**Sample usage, sign a URL:**

```javascript
const signedURL = signURL("user", "your-secret_key", "http://e.com/api/");
```

**Sample output:**

```javascript
'http://e.com/api/?valid_until=1378045287.0&auth_user=user&signature=YlZpLFsjUKBalL4x5trhkeEgqE8%3D'
```
**Options and defaults:**

The `signatureToDict` function accepts an optional `options` argument.

Default value for the `validUntil` in the `options` is 10 minutes from now. If
you want it to be different, set `validUntil` in the `options` of
the `signatureToDict` function.

Default lifetime of a signature is 10 minutes (600 seconds). If you want it
to be different, set `lifetime` in the `options` of the `signatureToDict`
function.

Default name of the (GET) param holding the generated signature value
is `signature`. If you want it to be different,set the `signatureParam`
in the `options` of the `signatureToDict` function.

Default name of the (GET) param holding the `authUser` value is
`auth_user`. If you want it to be different, set `authUserParam`
in the `options` of the `signatureToDict` function.

Default name of the (GET) param holding the `validUntil` value is
`valid_until`. If you want it to be different,  set the `validUntilParam`
in the `options` of the `signatureToDict` function.

Default name of the (GET) param holding the `extra` value is
`extra`. If you want it to be different, set the `extraParam`
in the `options` of the `signatureToDict` function.

Default hashing algorithm is `SHA1`. If you want it to be different, set the
`signatureCls` in the `options` of the `signatureToDict` function. Supported
classes are `HMACSHA1Signature` (alias of `Signature`), `HMACSHA256Signature`
and `HMACSHA512Signature`.

```javascript
signedData = signatureToDict(
    "user",
    "your-secret_key",
    {
        email: "john.doe@mail.example.com",
        first_name: "John",
        last_name: "Doe",
    },
    {
        authUserParam: "webshop_id",
    }
);
```

**Sample output:**

```javascript
{
    webshop_id: "user",
    email: "john.doe@mail.example.com",
    extra: "email,first_name,last_name",
    first_name: "John",
    last_name: "Doe",
    signature: "nu0Un+05z/cNOFnLwQnigoW/KmA=",
    valid_until: 1631799172.0
}
```

#### Recipient side

Validating the signed request data is as simple as follows.

##### Required imports

**CommonJS**

```javascript
const { validateSignedRequestData } = require("skajs");
```

**ESM**

```javascript
import { validateSignedRequestData } from "skajs";
```

##### Validate signed requests

Validating the signed request data. Note, that `data` value is expected to
be a dictionary; `request.POST` is given as an example.

```javascript
validationResult = validateSignedRequestData(
    request.POST, // Note, that ``request.POST`` is given as example.
    "your-secret_key"
);
```

In case of signed URLs, it could look as follows:

```javascript
validationResult = validateSignedRequestData(
    request.GET, // Note, that ``request.GET`` is given as example.
    "your-secret_key"
);
```

**Options and defaults:**

Similarly to `signatureToDict` function, the `validateSignedRequestData`
also accepts a number of optional arguments (which have been described above):

- signatureParam
- authUserParam
- validUntilParam
- extraParam
- signatureCls

With some customizations, it would look as follows:

```javascript
validationResult = validateSignedRequestData(
    request.GET,
    "your-secret_key",
    {
        authUserParam: "webshop_id",
    }
);
```

# Testing

Simply type:

```shell
npm test
```

# Code style

The `Prettier` is used.

```shell
npx prettier --write .
```

# License

MIT

# Support

For any issues contact me at the e-mail given in the [Author](#Author) section.

# Author

Artur Barseghyan <artur.barseghyan@gmail.com>


CHANGELOG.md
# Release history and notes
[Sequence based identifiers](http://en.wikipedia.org/wiki/Software_versioning#Sequence-based_identifiers)
are used for versioning (schema follows below):

```
major.minor[.revision]
```

- It's always safe to upgrade within the same minor version (for example, from
  0.3 to 0.3.4).
- Minor version changes might be backwards incompatible. Read the
  release notes carefully before upgrading (for example, when upgrading from
  0.3.4 to 0.4).
- All backwards incompatible changes are mentioned in this document.

## 0.1.3

2023-11-28

- Add `signURL` functionality.

## 0.1.2

2021-11-19

- Minor fixes.

## 0.1.1

2021-11-17

```text
Release dedicated to my dear mother.
```

- Added more signatures (`HMACSHA256Signature`, `HMACSHA512Signature`).

## 0.1.0

2021-11-13

- Nicer API.
- Fix unicode bug.

## 0.0.6

2021-08-20

- Initial public beta release.


examples.mjs
import {
    SIGNATURE_LIFETIME,
    getBase,
    sortedURLEncode,
    dictToOrderedDict,
    signatureToDict,
    generateSignature,
    makeHash,
    dictKeys,
    extractSignedData,
    validateSignature,
    makeValidUntil,
    validateSignedRequestData,
    DEFAULT_SIGNATURE_PARAM,
} from "./esm/index.js";

/**
 * Prepare a dictionary with data to sign.
 *
 * @param {Object} requestData
 * @returns {*}
 */
export function getSignatureData(requestData) {
    let signatureData = {};
    for (const [key, value] of Object.entries(requestData)) {
        if (SIGNATURE_DATA_KEYS.includes(key)) {
            signatureData[key] = value;
        }
    }
    return signatureData;
}

/**
 * Sample payload.
 */
const PAYLOAD = {
    order_lines: [
        {
            quantity: 4,
            product_id: "8273401260171",
            product_name: "himself",
            product_description: "Man movement another skill draw great late.",
            product_price_excl_tax: 7685,
            product_price_incl_tax: 8684,
            product_tax_rate_percentage: 13,
        },
        {
            quantity: 5,
            product_id: "6760122207575",
            product_name: "someone",
            product_description:
                "Including couple happen ago hotel son know list.",
            product_price_excl_tax: 19293,
            product_price_incl_tax: 20064,
            product_tax_rate_percentage: 4,
        },
        {
            quantity: 1,
            product_id: "5014352615527",
            product_name: "able",
            product_description: "Simply reason bring manager with lot.",
            product_price_excl_tax: 39538,
            product_price_incl_tax: 41910,
            product_tax_rate_percentage: 6,
        },
        {
            quantity: 1,
            product_id: "4666517682328",
            product_name: "person",
            product_description:
                "Arrive government such arm conference program every.",
            product_price_excl_tax: 18794,
            product_price_incl_tax: 18794,
            product_tax_rate_percentage: 0,
        },
        {
            quantity: 2,
            product_id: "3428396033957",
            product_name: "chance",
            product_description: "Ever campaign next store far stop and.",
            product_price_excl_tax: 26894,
            product_price_incl_tax: 29314,
            product_tax_rate_percentage: 9,
        },
        {
            quantity: 4,
            product_id: "4822589619741",
            product_name: "style",
            product_description: "Song any season pick box chance.",
            product_price_excl_tax: 17037,
            product_price_incl_tax: 19422,
            product_tax_rate_percentage: 14,
        },
    ],
    webshop_id: "4381a041-11cd-43fa-9fb4-c558bac1bd5e",
    order_id: "lTAGlTOHtKiBdvRvmhSw",
    amount: 491605,
    currency: "EUR",
    company: {
        name: "Siemens",
        registration_number: "LhkvLTWNTVNxlMKfBruq",
        vat_number: "RNQfPcPtnbDFvQRbJeNJ",
        website: "https://www.nedschroef.com/",
        country: "NL",
    },
    user: {
        first_name: "Noor",
        last_name: "van Praagh",
        email: "juliegoyaerts-van-waderle@gmail.com",
        phone_number: "+31475013353",
    },
    shipping: {
        street: "Femkeboulevard",
        house_number: "7",
        city: "Noord-Sleen",
        postal_code: "1784KL",
        country: "NL",
    },
    billing: {
        street: "Pippasteeg",
        house_number: "35",
        city: "Ospel",
        postal_code: "6385 VA",
        country: "NL",
    },
};

/**
 * *******************************************
 * *************** App specific **************
 * *******************************************
 */

/**
 * Shared secret
 */
const SECRET_KEY = "UxuhnPaO4vKA";

/**
 * Auth user.
 */
const AUTH_USER = "me@example.com";

/**
 * Fields to sign
 */
const SIGNATURE_DATA_KEYS = [
    "webshop_id",
    "order_id",
    "company",
    "order_lines",
    "amount",
    "currency",
    "user",
    "shipping",
    "billing",
];

/**
 * *******************************************
 * *************** Usage examples ************
 * *******************************************
 */

/**
 * Signature data
 */
const SIGNATURE_DATA = getSignatureData(PAYLOAD);

console.log("\n === \n signatureData \n === \n");
console.log(SIGNATURE_DATA);

const sortedSignatureData = dictToOrderedDict(SIGNATURE_DATA);

console.log("\n === \n sortedSignatureData \n === \n");
console.log(sortedSignatureData);

let validUntil = parseFloat(1628717009.0).toFixed(1);
console.log("\n === \n validUntil \n === \n");
console.log(validUntil);

const keys = dictKeys(PAYLOAD);

console.log("\n === \n keys \n === \n");
console.log(keys);

let hash = makeHash(
    AUTH_USER,
    SECRET_KEY,
    validUntil, // makeValidUntil(),
    sortedSignatureData
);

console.log("\n === \n hash \n === \n");
console.log(hash);

const hash2 = makeHash(
    AUTH_USER,
    SECRET_KEY,
    validUntil, // makeValidUntil(),
    null
);

console.log("\n === \n hash2 \n === \n");
console.log(hash2);

const hash3 = makeHash(
    AUTH_USER,
    SECRET_KEY,
    validUntil, // makeValidUntil(),
    { 1: "1", 2: "2" }
);

console.log("\n === \n hash3 \n === \n");
console.log(hash3);

// var validUntil = makeValidUntil();

// console.log(validUntil);

let signature = generateSignature(
    AUTH_USER,
    SECRET_KEY,
    validUntil,
    SIGNATURE_LIFETIME,
    null
);

console.log("\n === \n signature \n === \n");
console.log(signature);

const signature2 = generateSignature(
    AUTH_USER,
    SECRET_KEY,
    validUntil,
    SIGNATURE_LIFETIME,
    { 1: "1", 2: "2" }
);

console.log("\n === \n signature2 \n === \n");
console.log(signature2);

const signature3 = generateSignature(
    AUTH_USER,
    SECRET_KEY,
    validUntil,
    SIGNATURE_LIFETIME,
    SIGNATURE_DATA
);

console.log("\n === \n signature3 \n === \n");
console.log(signature3);

const signatureDict = signatureToDict(
    PAYLOAD["webshop_id"],
    SECRET_KEY,
    validUntil,
    SIGNATURE_LIFETIME,
    SIGNATURE_DATA,
    DEFAULT_SIGNATURE_PARAM,
    "webshop_id"
);

console.log("\n === \n signatureDict \n === \n");
console.log(signatureDict);

const signatureDict2 = signatureToDict(
    PAYLOAD["webshop_id"],
    SECRET_KEY,
    validUntil,
    SIGNATURE_LIFETIME,
    { 1: "1", 2: "2" },
    DEFAULT_SIGNATURE_PARAM,
    "webshop_id"
);

console.log("\n === \n signatureDict2 \n === \n");
console.log(signatureDict2);

const base = getBase(AUTH_USER, validUntil, null);

console.log("\n === \n base \n === \n");
console.log(base);

const base2 = getBase(AUTH_USER, validUntil, { 1: "1", 2: "2" });

console.log("\n === \n base2 \n === \n");
console.log(base2);

const encodedData = sortedURLEncode(SIGNATURE_DATA);
console.log("\n === \n encodedData \n === \n");
console.log(encodedData);

const encodedData2 = sortedURLEncode(SIGNATURE_DATA, false);
console.log("\n === \n encodedData2 \n === \n");
console.log(encodedData2);

const orderedPayload = dictToOrderedDict(PAYLOAD);
console.log("\n === \n orderedPayload \n === \n");
console.log(orderedPayload);

const extractedSignedData = extractSignedData(
    signatureDict,
    SIGNATURE_DATA_KEYS
);
console.log("\n === \n extractedSignedData \n === \n");
console.log(extractedSignedData);

let validSignature = generateSignature(
    AUTH_USER,
    SECRET_KEY,
    makeValidUntil(),
    SIGNATURE_LIFETIME,
    null
);
const isValidSignature = validateSignature(
    validSignature.signature,
    validSignature.authUser,
    SECRET_KEY,
    validSignature.validUntil,
    validSignature.extra
);
console.log("\n === \n isValidSignature \n === \n");
console.log(isValidSignature);

const validSignatureDict = signatureToDict(
    PAYLOAD["webshop_id"],
    SECRET_KEY,
    makeValidUntil(),
    SIGNATURE_LIFETIME,
    SIGNATURE_DATA,
    DEFAULT_SIGNATURE_PARAM,
    "webshop_id"
);
let payloadCopy = JSON.parse(JSON.stringify(PAYLOAD));
let updatedPayload = {
    ...payloadCopy,
    ...validSignatureDict,
};
const isValidRequestData = validateSignedRequestData(
    updatedPayload,
    SECRET_KEY,
    DEFAULT_SIGNATURE_PARAM,
    "webshop_id"
);
console.log("\n === \n isValidRequestData \n === \n");
console.log("\n === \n isValidRequestData::updatedPayload \n === \n");
console.log(updatedPayload);
console.log(isValidRequestData);


examples.js
const {
    SIGNATURE_LIFETIME,
    getBase,
    sortedURLEncode,
    dictToOrderedDict,
    signatureToDict,
    generateSignature,
    makeHash,
    dictKeys,
    extractSignedData,
    validateSignature,
    makeValidUntil,
    validateSignedRequestData,
    DEFAULT_SIGNATURE_PARAM,
} = require("skajs");

/**
 * Prepare a dictionary with data to sign.
 *
 * @param {Object} requestData
 * @returns {*}
 */
function getSignatureData(requestData) {
    let signatureData = {};
    for (const [key, value] of Object.entries(requestData)) {
        if (SIGNATURE_DATA_KEYS.includes(key)) {
            signatureData[key] = value;
        }
    }
    return signatureData;
}

/**
 * Sample payload.
 */
const PAYLOAD = {
    order_lines: [
        {
            quantity: 4,
            product_id: "8273401260171",
            product_name: "himself",
            product_description: "Man movement another skill draw great late.",
            product_price_excl_tax: 7685,
            product_price_incl_tax: 8684,
            product_tax_rate_percentage: 13,
        },
        {
            quantity: 5,
            product_id: "6760122207575",
            product_name: "someone",
            product_description:
                "Including couple happen ago hotel son know list.",
            product_price_excl_tax: 19293,
            product_price_incl_tax: 20064,
            product_tax_rate_percentage: 4,
        },
        {
            quantity: 1,
            product_id: "5014352615527",
            product_name: "able",
            product_description: "Simply reason bring manager with lot.",
            product_price_excl_tax: 39538,
            product_price_incl_tax: 41910,
            product_tax_rate_percentage: 6,
        },
        {
            quantity: 1,
            product_id: "4666517682328",
            product_name: "person",
            product_description:
                "Arrive government such arm conference program every.",
            product_price_excl_tax: 18794,
            product_price_incl_tax: 18794,
            product_tax_rate_percentage: 0,
        },
        {
            quantity: 2,
            product_id: "3428396033957",
            product_name: "chance",
            product_description: "Ever campaign next store far stop and.",
            product_price_excl_tax: 26894,
            product_price_incl_tax: 29314,
            product_tax_rate_percentage: 9,
        },
        {
            quantity: 4,
            product_id: "4822589619741",
            product_name: "style",
            product_description: "Song any season pick box chance.",
            product_price_excl_tax: 17037,
            product_price_incl_tax: 19422,
            product_tax_rate_percentage: 14,
        },
    ],
    webshop_id: "4381a041-11cd-43fa-9fb4-c558bac1bd5e",
    order_id: "lTAGlTOHtKiBdvRvmhSw",
    amount: 491605,
    currency: "EUR",
    company: {
        name: "Siemens",
        registration_number: "LhkvLTWNTVNxlMKfBruq",
        vat_number: "RNQfPcPtnbDFvQRbJeNJ",
        website: "https://www.nedschroef.com/",
        country: "NL",
    },
    user: {
        first_name: "Noor",
        last_name: "van Praagh",
        email: "juliegoyaerts-van-waderle@gmail.com",
        phone_number: "+31475013353",
    },
    shipping: {
        street: "Femkeboulevard",
        house_number: "7",
        city: "Noord-Sleen",
        postal_code: "1784KL",
        country: "NL",
    },
    billing: {
        street: "Pippasteeg",
        house_number: "35",
        city: "Ospel",
        postal_code: "6385 VA",
        country: "NL",
    },
};

/**
 * *******************************************
 * *************** App specific **************
 * *******************************************
 */

/**
 * Shared secret
 */
const SECRET_KEY = "UxuhnPaO4vKA";

/**
 * Auth user.
 */
const AUTH_USER = "me@example.com";

/**
 * Fields to sign
 */
const SIGNATURE_DATA_KEYS = [
    "webshop_id",
    "order_id",
    "company",
    "order_lines",
    "amount",
    "currency",
    "user",
    "shipping",
    "billing",
];

/**
 * *******************************************
 * *************** Usage examples ************
 * *******************************************
 */

/**
 * Signature data
 */
const SIGNATURE_DATA = getSignatureData(PAYLOAD);

console.log("\n === \n signatureData \n === \n");
console.log(SIGNATURE_DATA);

const sortedSignatureData = dictToOrderedDict(SIGNATURE_DATA);

console.log("\n === \n sortedSignatureData \n === \n");
console.log(sortedSignatureData);

let validUntil = parseFloat(1628717009.0).toFixed(1);
console.log("\n === \n validUntil \n === \n");
console.log(validUntil);

const keys = dictKeys(PAYLOAD);

console.log("\n === \n keys \n === \n");
console.log(keys);

let hash = makeHash(
    AUTH_USER,
    SECRET_KEY,
    validUntil, // makeValidUntil(),
    sortedSignatureData
);

console.log("\n === \n hash \n === \n");
console.log(hash);

const hash2 = makeHash(
    AUTH_USER,
    SECRET_KEY,
    validUntil, // makeValidUntil(),
    null
);

console.log("\n === \n hash2 \n === \n");
console.log(hash2);

const hash3 = makeHash(
    AUTH_USER,
    SECRET_KEY,
    validUntil, // makeValidUntil(),
    { 1: "1", 2: "2" }
);

console.log("\n === \n hash3 \n === \n");
console.log(hash3);

// var validUntil = makeValidUntil();

// console.log(validUntil);

let signature = generateSignature(
    AUTH_USER,
    SECRET_KEY,
    validUntil,
    SIGNATURE_LIFETIME,
    null
);

console.log("\n === \n signature \n === \n");
console.log(signature);

const signature2 = generateSignature(
    AUTH_USER,
    SECRET_KEY,
    validUntil,
    SIGNATURE_LIFETIME,
    { 1: "1", 2: "2" }
);

console.log("\n === \n signature2 \n === \n");
console.log(signature2);

const signature3 = generateSignature(
    AUTH_USER,
    SECRET_KEY,
    validUntil,
    SIGNATURE_LIFETIME,
    SIGNATURE_DATA
);

console.log("\n === \n signature3 \n === \n");
console.log(signature3);

const signatureDict = signatureToDict(
    PAYLOAD["webshop_id"],
    SECRET_KEY,
    SIGNATURE_DATA,
    {
        "validUntilParam": validUntil,
        "authUserParam": "webshop_id",
    }
);

console.log("\n === \n signatureDict \n === \n");
console.log(signatureDict);

const signatureDict2 = signatureToDict(
    PAYLOAD["webshop_id"],
    SECRET_KEY,
    { 1: "1", 2: "2" },
    {
        "validUntilParam": validUntil,
        "authUserParam": "webshop_id",
    }
);

console.log("\n === \n signatureDict2 \n === \n");
console.log(signatureDict2);

const base = getBase(AUTH_USER, validUntil, null);

console.log("\n === \n base \n === \n");
console.log(base);

const base2 = getBase(AUTH_USER, validUntil, { 1: "1", 2: "2" });

console.log("\n === \n base2 \n === \n");
console.log(base2);

const encodedData = sortedURLEncode(SIGNATURE_DATA);
console.log("\n === \n encodedData \n === \n");
console.log(encodedData);

const encodedData2 = sortedURLEncode(SIGNATURE_DATA, false);
console.log("\n === \n encodedData2 \n === \n");
console.log(encodedData2);

const orderedPayload = dictToOrderedDict(PAYLOAD);
console.log("\n === \n orderedPayload \n === \n");
console.log(orderedPayload);

const extractedSignedData = extractSignedData(
    signatureDict,
    SIGNATURE_DATA_KEYS
);
console.log("\n === \n extractedSignedData \n === \n");
console.log(extractedSignedData);

let validSignature = generateSignature(
    AUTH_USER,
    SECRET_KEY,
    makeValidUntil(),
    SIGNATURE_LIFETIME,
    null
);
const isValidSignature = validateSignature(
    validSignature.signature,
    validSignature.authUser,
    SECRET_KEY,
    validSignature.validUntil,
    validSignature.extra
);
console.log("\n === \n isValidSignature \n === \n");
console.log(isValidSignature);

const validSignatureDict = signatureToDict(
    PAYLOAD["webshop_id"],
    SECRET_KEY,
    SIGNATURE_DATA,
    {
        "authUserParam": "webshop_id",
    }
);
let payloadCopy = JSON.parse(JSON.stringify(PAYLOAD));
let updatedPayload = {
    ...payloadCopy,
    ...validSignatureDict,
};
const isValidRequestData = validateSignedRequestData(
    updatedPayload,
    SECRET_KEY,
    DEFAULT_SIGNATURE_PARAM,
    "webshop_id"
);
console.log("\n === \n isValidRequestData \n === \n");
console.log("\n === \n isValidRequestData::updatedPayload \n === \n");
console.log(updatedPayload);
console.log(isValidRequestData);
